src/evm/protocol/uniswap_v4/state.rs

```
@@ -280,6 +281,92 @@ impl ProtocolSim for UniswapV4State {
+                Err(_) => break,
+            };
+
+            let sqrt_price_next = get_sqrt_ratio_at_tick(next_tick)?;
+
+            let (amount_in, amount_out) = if zero_for_one {
+                let amount0 = get_amount0_delta(
+                    sqrt_price_next,
+                    current_sqrt_price,
+                    current_liquidity,
+                    true,
+                )?;
+                let amount1 = get_amount1_delta(
+                    sqrt_price_next,
+                    current_sqrt_price,
+                    current_liquidity,
+                    false,
+                )?;
+                (amount0, amount1)
+            } else {
+                let amount0 = get_amount0_delta(
+                    sqrt_price_next,
+                    current_sqrt_price,
+                    current_liquidity,
+                    false,
+                )?;
+                let amount1 = get_amount1_delta(
+                    sqrt_price_next,
+                    current_sqrt_price,
+                    current_liquidity,
+                    true,
+                )?;
+                (amount1, amount0)
+            };
+
+            total_amount_in = safe_add_u256(total_amount_in, amount_in)?;
+            total_amount_out = safe_add_u256(total_amount_out, amount_out)?;
+
+            if initialized {
+                let liquidity_raw = self
+                    .ticks
+                    .get_tick(next_tick)
+                    .unwrap()
+                    .net_liquidity;
+                let liquidity_delta = if zero_for_one { -liquidity_raw } else { liquidity_raw };
+                current_liquidity =
+                    liquidity_math::add_liquidity_delta(current_liquidity, liquidity_delta);
+            }
+
+            if zero_for_one && current_tick < next_tick {
+                break;
+            }
+
+            if !zero_for_one && current_tick == next_tick {
+                break;
+            }
```

I don't like this, I think `next_initialized_tick_within_one_word` should return an error when we're below the smallest tick instead of this:

```
            if self.is_below_smallest(tick) {
                let minimum = cmp::max(self.ticks[0].index - spacing, min_in_word);
                return Ok((minimum, false));
            }
```